src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.CheckOrReachability.cs

```diff
@@ -0,0 +1,1433 @@
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            public override BoundNode? VisitDiscardPattern(BoundDiscardPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            public override BoundNode? VisitRelationalPattern(BoundRelationalPattern node)
+            {
+                TryPushOperand(NegateIfNeeded(node));
+                return null;
+            }
+
+            // Given a pattern that expects `pattern.InputType` as input type, produce
+            // one that expects `inputType`.
+            //
+            // For example, when splitting a pattern `{ Prop: A and B }` into `{ Prop: A } and/or { Prop: B }`
+            // the original `B` pattern expects an input type that is the narrowed type from `A`
+            // but the rewritten `B` pattern expects an input type that is the type of `Prop`.
+            private static BoundPattern WithInputTypeCheckIfNeeded(BoundPattern pattern, TypeSymbol inputType)
+            {
+                if (pattern.InputType.Equals(inputType, TypeCompareKind.AllIgnoreOptions))
+                {
+                    return pattern;
+                }
+
+                if (pattern is BoundTypePattern typePattern1)
+                {
+                    return typePattern1.Update(typePattern1.DeclaredType, typePattern1.IsExplicitNotNullTest, inputType, typePattern1.NarrowedType);
+                }
+
+                if (pattern is BoundRecursivePattern recursivePattern)
+                {
+                    return recursivePattern.Update(
+                        recursivePattern.DeclaredType ??
+                            new BoundTypeExpression(recursivePattern.Syntax, aliasOpt: null, recursivePattern.InputType.StrippedType()),
+                        recursivePattern.DeconstructMethod, recursivePattern.Deconstruction,
+                        recursivePattern.Properties, recursivePattern.IsExplicitNotNullTest,
+                        recursivePattern.Variable, recursivePattern.VariableAccess,
+                        inputType, recursivePattern.NarrowedType);
+                }
+
+                if (pattern is BoundDiscardPattern discardPattern)
+                {
+                    return discardPattern.Update(inputType, inputType);
+                }
+
+                if (pattern is BoundNegatedPattern negatedPattern)
+                {
+                    return negatedPattern.Update(
+                        WithInputTypeCheckIfNeeded(negatedPattern.Negated, inputType),
+                        inputType, inputType);
+                }
+
+                if (pattern is BoundConstantPattern constantPattern)
+                {
+                    return constantPattern.Update(constantPattern.Value, constantPattern.ConstantValue, inputType, inputType);
```

Comment:
>inputType [](http://example.com/codeflow?start=115&length=9)

This looks suspicious. Perhaps this makes sense for a `null` pattern, but for other patterns it feels like narrowed type should not change.